### PR TITLE
Port mimetypes to tables

### PR DESCRIPTION
**Pull request by b3liever**
Port from stringtabs to tables module.

### CODE REVIEW

**Comment by Araq** [comment]
But why? Tying == and hash to string just for this module is very dangerous and will cause problems.

**Comment by b3liever** [comment]
Does it work across modules? I thought the overloads will only apply in the current module, no?

**Comment by b3liever** [comment]
Let run the CIs out of curiosity and then I have a patch that removes them, ready.

**Comment by Araq** [comment]
One problem is that the generic instantation cache doesn't even consider scoping, so a case sensitive Table[string, string] instantiation might be used instead. (Bad bug, I know, but also quite tough to fix and arguably ``==`` and ``hash`` should be attached to the type.)

**Comment by b3liever** [comment]
Thanks for explaining!

File: lib/pure/mimetypes.nim [hunk 1]
```diff
@@ -1916,7 +1912,7 @@ func register*(mimedb: var MimeDB, ext: string, mimetype: string) =
   assert mimetype.strip.len > 0, "mimetype argument can not be empty string"
   mimedb.mimes[ext.toLowerAscii()] = mimetype.toLowerAscii()
 
-runnableExamples:
+when isMainModule:
```
Comment (b3liever) [inline lib/pure/mimetypes.nim]: This removed the examples from the top of the page, sry will revert

File: lib/pure/mimetypes.nim [hunk 2]
```diff
@@ -8,12 +8,12 @@
...
-import strtabs
+import tables
 from strutils import startsWith, toLowerAscii, strip
 
 type
   MimeDB* = object
-    mimes: StringTableRef
+    mimes: Table[string, string]
 
 const mimes* = {
```
Comment (b3liever) [inline lib/pure/mimetypes.nim]: Shouldn't this be private field? Except polluting the documentation with many entries, it shouldn't be used directly

Comment (Araq) [reply lib/pure/mimetypes.nim]: Why not? It's more likely we got the ``const`` right than the API...

**Comment by b3liever** [comment]
This is ready, no?

**Comment by Araq** [comment]
But why not use ``newStringTable(modeCaseInsensitive)``? We should use data structures designed for the purpose.

**Comment by narimiran** [comment]
> But why not use `newStringTable(modeCaseInsensitive)`?

This basically returns us to the state from before this PR, doesn't it?
@b3liever, will you make these changes (by mostly reverting what you have done) or should we close this PR?

**Comment by b3liever** [comment]
Sorry for inactivity have forgotten this PR. I thought ``StringTableRef`` was legacy, that's why I tried to port  away from it. Currently ``mimetypes`` doesn't use it properly (it should call ``newStringTable(modeCaseInsensitive)`` as Araq pointed out). My concern is that it's a ``ref object`` while ``Table``, that I did the port to is a plain object and avoids an indirection.

But the most important issue is ``getExt`` proc. It does linear search on the table's values and this can't be avoided with either strtabs or tables. Thus I think that neither is actually a good fit for this module.

**Comment by Araq** [comment]
> But the most important issue is getExt proc. It does linear search on the table's values and this can't be avoided with either strtabs or tables. Thus I think that neither is actually a good fit for this module.

But we need to measure whether that is actually a problem.

**Comment by Araq** [comment]
There is little left in this PR, closing.

**Pull request closed without merging**